### ISSUE TITLE
Limbo support

### ISSUE THREAD

**Issue by jayvdb** [Electron100/butane#287]
https://avi.im/blag/2024/faster-sqlite/

SQLite compatible, built in Rust.

https://github.com/tursodatabase/limbo

Maybe it "just works" , or needs a separate backend, but it should be fairly minimal.

**Comment by RobertoMaurizzi** [comment]
Interesting especially because it'll have async support, but likely premature? :sweat_smile:

**Comment by jayvdb** [comment]
Ah, yes, missing UPDATE/UPSERT and DELETE makes this backend useless except for KV store like usage.

https://github.com/tursodatabase/limbo/blob/main/COMPAT.md

A https://github.com/tursodatabase/libsql/tree/main/libsql backend might be more useful atm.

**Comment by jayvdb** [comment]
It appears to be missing `UPSERT` still, which would require significant effort to workaround, but that effort is the same as is needed for https://github.com/Electron100/butane/issues/349 .

butane has code that can handle backends that dont support `ALTER TABLE`, so this isnt a blocker.

I think we could add a limbo backend even if rollback wasnt supported, as long as we informed butane users that it was not supported; it could even be `unimplemented!()` in the butane backend code.

**Comment by Electron100** [comment]
Do we have anybody interested in using it if we add it?

**Comment by Nekidev** [comment]
Yeah, I'm building an API that's gonna be self-hostable, but it should not die under high loads and SQLite with an async wrapper doesn't scale well for writes under high loads. Having this built in would be amazing because it'd solve the portability (which postgres doesn't give that seamlessly) AND the speed issue.

Still, this is a use case where I can still live with using postgres and nothing too bad would happen. It'd be useful if it were there but maybe it's not yet worth the maintainability cost of adding a third backend, specially when limbo docs still say at the top of their README that it's still NOT production ready.

**Comment by jayvdb** [comment]
We would use it instead of sqlite as the db used when running unit test , even if it is slower.

And replace sqlite as a supported mode, if it handles concurrent writes better, which I assume it will.

**Comment by jayvdb** [comment]
Mostly AI generated Turso support: https://github.com/Electron100/butane/pull/387

I'm seeing lots of failures, some due to missing sqlite compatibility, but some probably can be solved by finding workarounds.

**Comment by jayvdb** [comment]
There appears to primarily one SQL feature missing which is causing most of the failures, described in a .md file in the PR.

The PR is now green, with some tests marked "pg" to disable them on turso.

**Comment by jayvdb** [comment]
Turso support now merged.
